---
output: github_document
---


```{r, echo = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "README-"
)
```

# What is SetDesign?

SKATmultiallele was designed for power calculation of Sequece Kernal Association Test (SKAT)  when multi-allelic positions are present. Users can specify parameters such as the number of subjects, SNPs and the minor allele frequencies, significance level and effect size of both alternative alleles to compute the power of SKAT test. The package supports both linear and binary outcomes and accounts for correlated SNPs by specifying a correlation coefficient for more precise results.


# Why use SetDesign?

Traditional methods for SKAT power calculations often oversimplify multi-allelic positions by treating them as single allelesâ€”either considering only the first alternative allele or combining the first and second alternative alleles in a simplified manner. SKATmultiallele highlights the differences in power calculations when these positions are treated individually versus when they are combined.

Additionally, the package builds on and improves the power calculation method from Lee et al.'s paper, significantly reducing computational complexity by directly specifying the correlation structure of the genotype matrix.

# Example

We show an example with binary outcomes and correlated SNPs.

```{r k2, eval = T, echo=TRUE, warning = FALSE, message = FALSE, error = FALSE}
library(SKATmultiallele)

# Set number of  SNPs
kk = 50

# Set number of subjects
n = 2000

# Set significance level
alpha = 0.05

# Set correlation coefficient between SNPs
rho = 0.15

# Set minor allele frequency
p = 0.01

# Set effect sizes of the first alternative allele
list1=c(-0.1,-0.25,-0.5,-0.75,-0.9)

# Set effect sizes of the second alternative allele
list2=c(rep(0.5,5))

# Pairwise correlation
rho = 0.1

# get the power for SKAT test
powerD_derive(kk,n,p,alpha,list1,list2)
```
